### PR TITLE
refactored script_analyzers into analyzers_manager Django sub-app

### PR DESCRIPTION
**Pull request by m0mosenpai**
# Description

Refactored `script_analyzers` to it's own Django sub-app: `analyzers_manager`, following the directory structure mentioned in #397. Added basic code for the new `Analyzer`, `Secrets`, `AnalyzerReport` models.

## Related issues
#397 , #392 

## Type of change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# To-Do

- [x] Creating `analyzers_manager` sub-app and copying `script_analyzers` into it.
- [x] Replacing `script_analyzers` instances in analyzer files, tests and docs
- [x] Creating new Analyzer, Secret and AnalyzerReport models
- [x] Incorporating `Analyzer` model in filtering, executing jobs.
- [x] Changes in how reports are generated (utils.py, classes.py)
- [x] Registering models in Admin interface
- [ ] Database Migrations for analyzer_config.json
- [ ] Refactoring everything (cleaner code, adding type information, docstrings etc.)

# Checklist
- [X] The pull request is for the branch develop
- [ ] I have added tests in the [Tests](https://github.com/intelowlproject/IntelOwl/blob/master/tests) folder. 
- [ ] The tests gave 0 errors.
- [x] `Black` gave 0 errors.
- [x] `Flake` gave 0 errors.
- [ ] I squashed the commits into a single one. (optional, they will be squashed anyway by the maintainer)

# Real World Example

TBD

### CODE REVIEW

**Comment by m0mosenpai** [comment]
I have made a preliminary PR for refactoring and adding mostly the code @Eshaan7 had written before along with some more additions. I haven't removed existing `script_analyzers` yet. Was thinking if the directory structure is correct, I will continue to build on it and remove the old folder. (this way i won't break functionality immediately).

**Comment by mlodic** [comment]
are all of these new files? It is important to keep the history while moving the files.

The python files should be moved, so we cannot keep both versions.
This is an example of what we should see from Github:
image

It is also very difficult to understand what was changed in this way because Github understand that everything is new in this way.

This is a developing branch so do not be afraid of breaking functionality. We will merge this on `develop` only when all the work will be completed

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:12]: can you extract out the `choices` into a constant variable inside the `Analyzer` model ? thanks

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:79]: extract out into a constant variable `CHOICES` inside Secret class

File: api_app/analyzers_manager/views.py
Comment (Eshaan7) [inline api_app/analyzers_manager/views.py:8]: use `viewsets.ReadOnlyModelViewSet` because we do not wish to allow update/create at the moment.

File: intel_owl/settings.py
Comment (Eshaan7) [inline intel_owl/settings.py:53]: Since the `api_app` is a django app and the `analyzers_manager` only a sub-app, we shouldn't add it here.

This also means no need for the `analyzers_manager/apps.py` file.

Same is true for `connectors_manager`, so ping @sp35.

File: api_app/analyzers_manager/admin.py
Comment (Eshaan7) [inline api_app/analyzers_manager/admin.py:3]: I guess it is already in your to-do list, but we would need admin interfaces against all the new models here.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:44]: isn't this condition wrong ? if there any `missing_secrets`, that means `configured` should be False.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:28]: remove this. it's broken anyway.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:40]: replace with `self.__cached_secrets[secret.name]` note the extra underscore.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:47]: using `self.secrets.all().count()` is more performant and django-recommended way.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:64]: add `default=False`.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:71]: remove the started_time_str field, it's pretty redundant.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:70]: should be a `models.DateTimeField`

Comment (Eshaan7) [reply api_app/analyzers_manager/models.py:70]: add a `end_time` field too

File: intel_owl/settings.py
Comment (sp35) [reply intel_owl/settings.py:53]: How would django recognize this sub app then (e.g. for making migrations)?

Comment (m0mosenpai) [reply intel_owl/settings.py:53]: Even I made this change after finding that a sub-app is independent from the main app and would require it to be added in the settings.py separately. I'll test it out without this line as well once

File: api_app/analyzers_manager/models.py
Comment (m0mosenpai) [reply api_app/analyzers_manager/models.py:44]: yeah, my bad

File: api_app/analyzers_manager/models.py
Comment (sp35) [inline api_app/analyzers_manager/models.py:85]: This may be done using a through model `AnalyzerSecret` with one to one fields of `Analyzer` and `Secret` and correspondingly `ConnectorSecret`. This way I can use the `Secret` model for `Connector` also.

File: api_app/analyzers_manager/utils.py
Comment (m0mosenpai) [inline api_app/analyzers_manager/utils.py:77]: I assume, I will also need to pass in `Analyzer` and `Job` model objects (since `AnalyzerModel` has a foreign key relationship with them? If so, for `Job`, i'll need to do some refactoring but i was confused how to incorporate the `Analyzer` model here.

File: intel_owl/settings.py
Comment (Eshaan7) [reply intel_owl/settings.py:53]: AFAIK and I have used it like this before, just adding the "api_app" to the list should be sufficient. Django recursively traverses all sub-directories.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [reply api_app/analyzers_manager/models.py:85]: ^ yep, definetly correct.

File: api_app/analyzers_manager/utils.py
Comment (Eshaan7) [reply api_app/analyzers_manager/utils.py:77]: This function shouldn't be inside the `utils.py`. Remember that we are refactoring the codebase to make it better so we don't necessarily need to adopt previous code. 

A better approach inline with new changes would be to have a member method `init_report` inside `Analyzer` model. The signature being something like: `def init_report(self) -> AnalyzerReport`

**Comment by Eshaan7** [comment]
Could you include a todo type list in this PR ? like create a checkbox list of everything that you aim to do in this PR and keep crossing them off as you progress ? that way it will also be easier for us to track progress. Thanks.

File: api_app/analyzers_manager/utils.py
Comment (Eshaan7) [reply api_app/analyzers_manager/utils.py:77]: Another way could be to have this method inside `Job` model. `def init_report(self, analyzer: Analyzer) -> AnalyzerReport`. I'll let you decide which one is better as you are coding it.

File: api_app/analyzers_manager/classes.py
Comment (Eshaan7) [inline api_app/analyzers_manager/classes.py:4]: `from django.utils import timezone` and then use `timezone.now()`. Otherwise django throws naive datetime warnings.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [reply api_app/analyzers_manager/models.py:85]: To enable re-usability of `Secret` model, it should be placed inside a different `models.py` to avoid circular import issues.

File: intel_owl/settings.py
Comment (Eshaan7) [reply intel_owl/settings.py:53]: @m0mosenpai when you ran `makemigrations`, did it create migration files inside `api_app/migrations` or `api_app/analyzers_manager/migrations/` ?

Comment (m0mosenpai) [reply intel_owl/settings.py:53]: Yeah, i got a `0001_initial.py` in `analyzers_manager/migrations` and a `0005_auto_xxxx.py` in `api_app/migrations`

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:61]: Instead of a boolean field like this, i think better would be to have a `status = models.CharField(choices=...)`, with choices like: `pending`, `running`, `failed`, `success`.

File: intel_owl/settings.py
Comment (Eshaan7) [reply intel_owl/settings.py:53]: okay then that's good so we should keep the line including the sub-app in `INSTALLED_APPS`.

File: api_app/analyzers_manager/utils.py
Comment (m0mosenpai) [reply api_app/analyzers_manager/utils.py]: I put it in Job cause the report model as i understood, represents a Job's report (which may contain individual reports from different analyzers) so it made more sense to me that way

File: api_app/analyzers_manager/models.py
Comment (sp35) [inline api_app/analyzers_manager/models.py:20]: can be skipped: `blank` and `null` are `False` by default

Comment (sp35) [reply api_app/analyzers_manager/models.py:20]: Please omit this everywhere!

**Review by sp35 — CHANGES_REQUESTED** [review]

File: api_app/analyzers_manager/models.py
Comment (sp35) [inline api_app/analyzers_manager/models.py:49]: I feel there is no use in keeping `blank` and `null` when you have set a default value.

File: api_app/analyzers_manager/models.py
Comment (sp35) [inline api_app/analyzers_manager/models.py:48]: Maybe use `PositiveIntegerField`?

File: api_app/analyzers_manager/models.py
Comment (m0mosenpai) [reply api_app/analyzers_manager/models.py]: I have changed where I have used `blank` and `null` based on some documentation I read. There were places it was not good of me to use them, so removed from there. But otherwise, i would prefer to have `blank` at other places for the sake of being a bit explicit.

File: api_app/analyzers_manager/models.py
Comment (Eshaan7) [inline api_app/analyzers_manager/models.py:108]: use these same fields in the `ConnectorReport`.

**Pull request closed without merging**
